### PR TITLE
Fix tests setup - ensure errors are exposed

### PR DESCRIPTION
**Pull request by medikoo**
## What did you implement:

There are errors in tests which are not properly exposed, additionally due to that, some configured tests are not invoked at all.

Issue lies in not handled rejected promises, their errors are exposed only via _UnhandledRejection_ warnings which can be seen through the tests output, but which do not force tests to fail as expected.

I didn't fix code corresponding to broken tests, I thought it's better (more efficient) if it's looked after by its authors. __Due to that CI crashes, it's expected__

## How did you implement it:

In tests I ensured that those rejected promises are returned to mocha functions, therefore exposed as expected.

Still, more future proof solution would be to add following code to some init test script:

```javascript
process.on('unhandledRejection', err => {
  throw err;
});
```

This will ensure that all currently (and those introduced in future) broken tests  will fail same way as they do after this PR is applied.

This approach was also scheduled to be default in Node.js -> https://github.com/nodejs/node/pull/12010
If you agree on that solution I may add it instead of patching individual tests, just let me know where's the best place.

## How can we verify it:

Run tests

## Todos:

***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO

### CODE REVIEW

**Comment by pmuens** [comment]
Looks like we already have smth. like that in place --> [here](https://github.com/serverless/serverless/blob/0d177a4bc642c11e98848791180cc548f6bf60be/bin/serverless#L17-L19).

It will execute [this method](https://github.com/serverless/serverless/blob/0d177a4bc642c11e98848791180cc548f6bf60be/lib/classes/Error.js#L41-L96). However it's not executed correctly (looks like the exit code is not set correctly).

**Comment by medikoo** [comment]
> Looks like we already have smth. like that in place --> here.

This program is not executed (required) when testing, which probably makes sense, but then it means that `process.on('unhandledRejection',...` should be introduced also in some module that's invokved by tests, which one would be right?

**Comment by HyperBrain** [comment]
@pmuens The problem with invoke local is, that customer code is executed within the Serverless process. That means that, if the frameworks adds a specific unhandled rejection or unhandled exception handler, it must make sure that the tracker is not invoked in case of `invoke local` to prevent sending sensitive data.

The other side of this is, that the unhandled handler should in this case not interfere with handlers that the user code might have set. Imagine user code that handles them in some obscure way and just continues execution somehow.

Maybe the handler could be set conditionally in the framework, depending on the commands invoked (i.e. for `invoke local`) the handler should not be set. Would this be a feasible solution, or did I miss something?

**Comment by medikoo** [comment]
@HyperBrain do you points also relate to tests process in any way?

I totally agree it should not be forced within Serverless in general, but there's no side effects to the users of the framework if unhandled rejections are forced to crash purely in program that runs tests (or am I missing something?)

**Comment by HyperBrain** [comment]
I referred to @pmuens 
> should be introduced also in some module that's invoked by tests, 

Locally within the tests I see no problem at all. Only if the handling leaves that scope care has to be taken 😃

**Comment by medikoo** [comment]
Ideally this should be taken care in mocha, and it's already been proposed and discussed there:

https://github.com/mochajs/mocha/issues/1926
https://github.com/mochajs/mocha/issues/2640 (They even seem to be open for PR)

For a meantime we may take care of it via slight hacking, introduce `bin/test` with following content:

```javascript
#!/usr/bin/env node

'use strict';

process.on('unhandledRejection', err => {
  throw err;
});

require('mocha/bin/_mocha');
```

And referring scripts to this program instead of `node_modules/bin/_mocha`

**Comment by pmuens** [comment]
Thanks for the nice conversation @HyperBrain and @medikoo 👍!

I like the solution to introduce a `bin/test` file and "hack" around the problem for now since this won't break any existing UX.

**Comment by medikoo** [comment]
@pmuens I thought it might be good to also hard code here arguments:  

`\"!(node_modules)/**/*.test.js\" --require=sinon-bluebird -R spec --recursive`

 It can work as default, so if some arguments are detected, then they're not forced

It will allow, to not have them duplicated in [scripts configuration](https://github.com/serverless/serverless/blob/master/package.json#L54-L55)

What do you think?

**Comment by pmuens** [comment]
@medikoo yes, that sounds like a good enhancement! This way we'd have one place to look for / make changes 👍

**Comment by medikoo** [comment]
@pmuens I applied that. I've also added `--no-exit` to mocha setup, as without this, tests program would still be vulnerable to hidden errors issue, more info on that here -> https://github.com/mochajs/mocha/issues/2879

**Comment by pmuens** [comment]
> @pmuens I applied that. I've also added --no-exit to mocha setup, as without this, tests program would still be vulnerable to hidden errors issue, more info on that here -> mochajs/mocha#2879

Very nice! Thanks for the updates @medikoo 👍!

File: lib/Serverless.test.js [hunk 1]
```diff
@@ -207,9 +206,10 @@ describe('Serverless', () => {
         .stub(serverless.variables, 'populateService').resolves();
       runStub = sinon
         .stub(serverless.pluginManager, 'run').resolves();
-    });
+    }));
 
     afterEach(() => {
+      if (!serverless.utils.logStat.restore) return;
```
Comment (pmuens) [inline lib/Serverless.test.js]: Why is this needed?

File: lib/classes/Utils.test.js [hunk 2]
```diff
@@ -21,7 +21,7 @@ describe('Utils', () => {
   beforeEach(() => {
     serverless = new Serverless();
     utils = new Utils(serverless);
-    serverless.init();
+    return serverless.init();
```
Comment (pmuens) [inline lib/classes/Utils.test.js]: Shouldn't we put this right at the beginning of the `beforeEach` block like it's done below?

File: lib/classes/Utils.test.js [hunk 3]
```diff
@@ -305,18 +305,17 @@ describe('Utils', () => {
...
       serverless.processedInput.commands = [];
       serverless.processedInput.options = {};
 
       trackStub = sinon.stub(segment, 'track');
       getConfigStub = sinon.stub(configUtils, 'getConfig');
-    });
+    }));
 
     afterEach(() => {
+      if (!segment.track.restore) return;
```
Comment (pmuens) [inline lib/classes/Utils.test.js]: Why is this needed?

File: lib/utils/fs/writeFile.test.js [hunk 4]
```diff
@@ -16,7 +16,7 @@ describe('#writeFile()', function () {
   this.timeout(0);
   beforeEach(() => {
     serverless = new Serverless();
-    serverless.init();
+    return serverless.init();
```
Comment (pmuens) [inline lib/utils/fs/writeFile.test.js]: Shouldn't we move this to the beginning of the `beforeEach` block?

File: lib/utils/fs/writeFileSync.test.js [hunk 5]
```diff
@@ -11,7 +11,7 @@ describe('#writeFileSync()', () => {
 
   beforeEach(() => {
     serverless = new Serverless();
-    serverless.init();
+    return serverless.init();
```
Comment (pmuens) [inline lib/utils/fs/writeFileSync.test.js]: Shouldn't we move this to the beginning of the `beforeEach` block?

File: lib/Serverless.test.js (hunk 1 above)
Comment (medikoo) [reply lib/Serverless.test.js]: As `serverless.init()` fails, stubs are not initialized therefore there's no `restore` methods.
Without this, there will be additional crash in `afterEach`.

Still it's indeed not needed when `serverless.init()` succeeds, so assuming we're fix it, we can indeed remove it, as it's just extra noise

File: lib/classes/Utils.test.js (hunk 2 above)
Comment (medikoo) [reply lib/classes/Utils.test.js]: You mean to move it before `utils = new Utils(serverless);` line? It can't be starting line here, as there's no `serverless` instance yet

File: lib/classes/Utils.test.js (hunk 3 above)
Comment (medikoo) [reply lib/classes/Utils.test.js]: Same case as commented above

File: lib/utils/fs/writeFile.test.js (hunk 4 above)
Comment (medikoo) [reply lib/utils/fs/writeFile.test.js]: Here similarly, at beginning of this `beforeEach` block, there's no `serverless` instance

**Comment by medikoo** [comment]
> Furthermore I added some comments while doing a code-review.

@pmuens Thanks, I replied comments

> When I run it I get a pretty weird output. The result is duplicated. I've attached a screenshot which shows how it looks like on my machine after running npm test:

This is result of another unhandled rejection that got in after I proposed this PR (must came in after merge with master).

I fixed those with https://github.com/serverless/serverless/pull/4139/commits/d6b29ab33884f0bbc379b326804f814e72479c0e

Now, concerning weird output by mocha. The crashing now _unhandled rejections_ are picked by `mocha` as it observes internally `uncaughtException` events. Still the _mocha_ reports generated for that errors are not that reliable.
I propose then to force `unanhandledRejections` to crash process the hard way  (so they're not picked by mocha), then the error message is eminent and should leave no doubt, e.g. in this case:

screen shot 2017-08-25 at 11 36 59

It's actually how I picked all of them cleanly, for this PR. We can achieve it by, clearing all `uncaughtException` listeners prior throwing exception:

```javascript
process.on('unhandledRejection', err => {
  delete process._events.uncaughtException;
  throw err;
});
```

**Comment by medikoo** [comment]
@pmuens interesting, I do not observe it on my side.

I've also merged it with master and pursued repository cleanup with `git clean -dfx`, to ensure any cache does not go in place, and still have clean exit with no timeout but 10 tests failing. (I'm on OSX v10.12.6, node: v8.4.0, npm: v4.6.1)

Do you have some changes that are maybe not pushed into this branch? Also maybe try `git clean -dfx` to reset repository (?)

**Comment by pmuens** [comment]
Hey @medikoo thanks for the reply and thanks for updating this PR :+1:

The failing tests win my local machine are the same which are also failing on Travis (https://travis-ci.org/serverless/serverless/jobs/270768136#L2970). I tried to dig deeper into that (used `it.only` to only run e.g. the `should load serverless.yaml from filesystem` test) and it turns out that the tests timeout when I only run one of them, so it looks like there's something off with the Promise resolving 🤔

File: bin/test [hunk 6]
```diff
@@ -0,0 +1,22 @@
+#!/usr/bin/env node
+
+'use strict';
+
+process.on('unhandledRejection', err => {
```
Comment (HyperBrain) [inline bin/test]: I think this is wrong for the test suite.

The problem here is, that it prevents `chai` from testing rejections correctly. The tests and the run code must not set any `unhandledRejection` handler - more exactly, it must not turn rejections into exceptions.

It is very important that `chai-as-promised` continues after receiving an exception, because for `.to.be.rejectedWith()` tests the rejection will not be detected anymore by chai. This fails currently:
```
return expect(myFunctionThatShouldReject())
  .to.be.rejectedWith(/my error message/);
```

The rejection test is never reached, because someone forces the current test to be terminated.

Comment (medikoo) [reply bin/test]: @HyperBrain this is about _unhandled_ promise rejections, so ones _not handled_, either by `chai-as-promised` or `mocha` or anything else.

What's in your example is promise that's obviously handled (passed to `expect` which via `rejectedWith` is handled by `chai-as-promised`), you can easily validate that by e.g. adding test as:

```javascript
  it.only('test', () =>
    expect(Promise.reject(new Error('Rejection'))).to.be.rejectedWith('Rejection'));
```
It'll pass as expected, and it won't be emitted as `unhandledRejection`

Example of _unhandled rejection_  would be a promise _not handled by any entity_, so: not passed to any promise processing function and not returned to any promise processing consumer, as e.g.:
```javascript
 it.only('test', () => {
  Promise.reject(new Error('Rejection'));
});
```

**Comment by medikoo** [comment]
>  I tried to dig deeper into that (used it.only to only run e.g. the should load serverless.yaml from filesystem test) and it turns out that the tests timeout when I only run one of them, so it looks like there's something off with the Promise resolving 🤔

@pmuens  It's interesting, as again I cannot reproduce that. Can you post diff of your changes against this branch?

In my case when I have `it.only` just on `should load serverless.yaml from filesystem` (so just this single test is run. It passes (and no timeout) all is fine.

I actually need to add `it.only` to prior test: `should load serverless.yml from filesystem` to have the following  `should load serverless.yaml from filesystem` test fail. Still it happens with no timeout you observe

So e.g. after applying following diff:

```diff
diff --git a/lib/classes/Service.test.js b/lib/classes/Service.test.js
index b101845c..3f0abbfa 100644
--- a/lib/classes/Service.test.js
+++ b/lib/classes/Service.test.js
@@ -123,7 +123,7 @@ describe('Service', () => {
       return expect(noService.load()).to.eventually.resolve;
     });
 
-    it('should load serverless.yml from filesystem', () => {
+    it.only('should load serverless.yml from filesystem', () => {
       const SUtils = new Utils();
       const serverlessYml = {
         service: 'new-service',
@@ -181,7 +181,7 @@ describe('Service', () => {
       });
     });
 
-    it('should load serverless.yaml from filesystem', () => {
+    it.only('should load serverless.yaml from filesystem', () => {
       const SUtils = new Utils();
       const serverlessYml = {
         service: 'new-service',
```

I have test output as:

screen shot 2017-09-01 at 14 48 59

I run `npm run test-bare`, but exactly same way test fails with _coverage_ overhead when running `npm test`

**Comment by pmuens** [comment]
Thanks for taking a deep dive into that @medikoo 👍 💯

I'm running your branch with `0` changes. I too get the error with the missing Plugin, but only when I run multiple tests. If I only run on with `it.only` I get a timeout 🤔.

BTW. I'm running it on Linux (in Docker).

**Comment by HyperBrain** [comment]
@medikoo @pmuens I did a `git bisect` over the range [`1.12.1`, `master] to find the commit that introduced the timeouts in the unit tests. The bisect ended up with the following commit:
```
git.exe bisect bad

7664c2dfedfb9c529321dcd3584c70a7dd74188b is the first bad commit
commit 7664c2dfedfb9c529321dcd3584c70a7dd74188b
Author: Alex DeBrie <alexdebrie1@gmail.com>
Date:   Mon Aug 14 22:05:23 2017 +0000

Add FrameworkId to stats

:040000 040000 2e1249a87866d36cfe0a6dba9e0ee4add397fb23 d0a79267a10c8e609de5ea3d4496b668e8a2c651 M    lib

Success (813 ms @ 9/2/2017 12:11:58 AM)
```
I verified that with the commit immediately before, the unit tests still worked on my local Ubuntu, and this commit leads to timeouts for nearly all file operation tests.

@pmuens identified the PR containing the commit as: https://github.com/serverless/serverless/pull/4087

Imo we should analyze it and fix this issue first, before continuing to apply further (and maybe not really needed) fixed to the unit tests, so that we really improve them.

**Comment by pmuens** [comment]
Thanks for taking the time to bisect this @HyperBrain 👍

> Imo we should analyze it and fix this issue first, before continuing to apply further (and maybe not really needed) fixes to the unit tests, so that we really improve them.

Yes, I 100% agree. Let's fix this first and then continue (if necessary). I'll take a look into this today. Would be super nice if we could get the tests fixed ASAP so that we can merge this and have a reliable test suite again!

**Comment by pmuens** [comment]
So I think I figured out what broke the tests here.

In the tests for `Service.test.js` we're calling `init()` (from `lib/Serverless.js`) for some weird reason. This results in a problem with the function which generates the cache file for the Plugins in the `PluginManager` class (usually we'd stub this out). I've removed the unnecessary call to `init()` and the tight coupling here. Now only the `Service` related functionality is tested in isolation.

Some tests are still failing. I'm looking into those right now...

**Comment by medikoo** [comment]
@pmuens it's quite puzzling I'm not able to reproduce the timeout issue on my setup. 

You say you only experience it when you add `.only` to some test (so only this test is run). Can you post me a diff against this branch, which I can apply to exactly reproduce the setup that ends with timeout on your side?

Also on which version of node.js and linux you're observing that? I assume that it's npm v5 with dependencies installed up to `package-lock.json`?

**Comment by pmuens** [comment]
@medikoo thanks for you comment (saw it now after reading through my E-Mails).

🤔 that's super odd. @HyperBrain also faced the timeout issues. Maybe it was a race condition? I've fixed all the tests in the meantime and they should be quite stable right now.

Thanks for for getting this started and your work on this PR @medikoo 👍 💪
